hail/src/main/scala/is/hail/cxx/PackDecoder.scala

```diff
@@ -183,20 +183,33 @@ object PackDecoder {
 
     decoderBuilder += s"${ decoderBuilder.name }(std::shared_ptr<InputStream> is) : $buf(std::make_shared<$bufType>(is)) { }"
 
-    val rowFB = decoderBuilder.buildMethod("decode_row", Array("Region *" -> "region"), "char *", const = true)
-    val region = rowFB.getArg(0)
-    val initialSize = rt match {
-      case _: PArray | _: PBinary => 8
-      case _ => rt.byteSize
+    rt.fundamentalType match {
```

Comment:
This could be factored a bit better. Something like
```scala
    val (resType, initialSize) = rt.fundamentalType match {
      case node if node.isPrimitive => (typeToCXXType(rt), rt.byteSize)
      case _: PArray | _: PBinary => ("char *", 8)
      case _ => ("char *", rt.byteSize)
    }
    val rowFB = decoderBuilder.buildMethod("decode_row", Array("Region *" -> "region"), resType, const = true)
    val region = rowFB.getArg(0)
    val row = rowFB.variable("row", "char *", s"$region->allocate(${ rt.alignment }, $initialSize)")
    rowFB += row.define
    rowFB += decode(t.fundamentalType, rt.fundamentalType, buf.ref, region.ref, row.ref, rowFB)
    rowFB += (rt.fundamentalType match {
      case node if node.isPrimitive => s"return *reinterpret_cast<${typeToCXXType(rt)} *>($row);"
      case _: PArray | _: PBinary => s"return load_address($row);"
      case _ => s"return $row;"
    })
    rowFB.end()
```
avoids a lot of duplicated code.